platform/android/src/style/sources/custom_geometry_source.cpp

```diff
@@ -102,20 +151,38 @@ namespace android {
     void CustomGeometrySource::registerNative(jni::JNIEnv& env) {
         // Lookup the class
         CustomGeometrySource::javaClass = *jni::Class<CustomGeometrySource>::Find(env).NewGlobalRef(env).release();
+        static auto nativePtrField = javaClass.GetField<jlong>(env, "nativePtr");
 
         #define METHOD(MethodPtr, name) jni::MakeNativePeerMethod<decltype(MethodPtr), (MethodPtr)>(name)
 
-        // Register the peer
+        // Register the peer instance methods
         jni::RegisterNativePeer<CustomGeometrySource>(
             env, CustomGeometrySource::javaClass, "nativePtr",
-            std::make_unique<CustomGeometrySource, JNIEnv&, jni::Object<CustomGeometrySource>, jni::String, jni::Object<>>,
-            "initialize",
-            "finalize",
             METHOD(&CustomGeometrySource::querySourceFeatures, "querySourceFeatures"),
             METHOD(&CustomGeometrySource::setTileData, "nativeSetTileData"),
             METHOD(&CustomGeometrySource::invalidateTile, "nativeInvalidateTile"),
             METHOD(&CustomGeometrySource::invalidateBounds, "nativeInvalidateBounds")
         );
+
+        auto initialize = [](jni::JNIEnv& env, jni::Object<CustomGeometrySource> javaSource, jni::String sourceId,
+                             jni::Object<> options) {
+            auto sharedSource = std::make_shared<CustomGeometrySource>(env, javaSource, sourceId, options);
+            javaSource.Set(env, nativePtrField, reinterpret_cast<jni::jlong>(sharedSource.get()));
+            sharedSource->source.peer = SourceWrapper { sharedSource->shared_from_this() };
+        };
+
+        auto finalize = [](jni::JNIEnv& env, jni::Object<CustomGeometrySource> javaSource) {
+            auto sharedSource = std::shared_ptr<CustomGeometrySource>(reinterpret_cast<CustomGeometrySource *>((jlong)javaSource.Get(env, nativePtrField)));
+            if (sharedSource) javaSource.Set(env, nativePtrField, (jlong)0);
+            sharedSource.reset();
```

Comment:
This resets only the local instance of `shared_ptr`, not the one in the C++ core `peer` field, so both the C++ peer and the C++ core object will leak when a `CustomGeometrySource` that's not added to a map is finalized.